### DISCUSSION TITLE
Anchor part vs instance cropping center?

### DISCUSSION THREAD

**Discussion by sfmig** [talmolab/sleap#1415] [Help!] [answered]
Hi,

I was reviewing the training job config parameters and realised that there are a couple of parameters that in my mind relate to the same the anchor point. 

Using the syntax for the TrainingJobConfig class (with `cfg = sleap.nn.config.TrainingJobConfig()`) we have both: 
```
cfg.data.instance_cropping.center_on_part
```
and 
```
cfg.model.heads.centroid.anchor_part
cfg.model.heads.centered_instance.anchor_part
```
for the top-down model stages.

I was wondering why these are separately defined and if there would be a case in which they could be different bodyparts. Are they forced to match? If so, which one takes priority?

Thanks!

**Comment by talmo** [comment, marked as answer]
Hi @sfmig,

Thanks for reaching out!

These do indeed refer to the same thing. They should point to the same anchor part to be consistent, and we enforce that when you generate the config via the GUI.

The reason we keep them separate is because they feed into different components of the modular backend. They don't really take priority over each other, they just affect different components. Specifically:

- `cfg.data.instance_cropping.center_on_part` sets up the [`InstanceCentroidFinder`](https://github.com/talmolab/sleap/blob/f9d0a2205d7d0f735639f5e9f89e59aab54b5958/sleap/nn/data/instance_centroids.py#L54-L195) and [`InstanceCropper`](https://github.com/talmolab/sleap/blob/f9d0a2205d7d0f735639f5e9f89e59aab54b5958/sleap/nn/data/instance_cropping.py#L224-L458) blocks of the data pipeline so that it generates training data with appropriate centering for top-down models.
- `cfg.model.heads.centroid.anchor_part` configures the [`CentroidConfmapsHead`](https://github.com/talmolab/sleap/blob/f9d0a2205d7d0f735639f5e9f89e59aab54b5958/sleap/nn/heads.py#L116C1-L154)
- `cfg.model.heads.centered_instance.anchor_part` configures the [`CenteredInstanceConfmapsHead`](https://github.com/talmolab/sleap/blob/f9d0a2205d7d0f735639f5e9f89e59aab54b5958/sleap/nn/heads.py#L157-L209)

These are also used as metadata for inference so we know how to parse model outputs into predictions.

There's not really a use case for changing them separately for most people. We do try to enforce consistency where possible, but technically you can change them individually when working with the low-level API. From the engineering perspective, one advantage to allowing them to be controlled independently is so that we can [test the components individually](https://github.com/talmolab/sleap/blob/f9d0a2205d7d0f735639f5e9f89e59aab54b5958/tests/nn/data/test_instance_cropping.py#L66-L169). Another is that it reduces [coupling](https://en.wikipedia.org/wiki/Loose_coupling) across components -- in this case, the model heads don't depend on how the data pipeline specifies the anchor part.

Let us know if you have any questions!

Talmo

**Comment by sfmig** [comment]
fantastic, thanks for the detailed answer! ✨ 
It's great to understand a bit better the software engineering decisions behind the design of the package
